--- a/net/wireless/scan.c
+++ b/net/wireless/scan.c
@@ -1365,7 +1365,7 @@ void cfg80211_bss_age(struct cfg80211_re
                       unsigned long age_secs)
 {
 	struct cfg80211_internal_bss *bss;
-	unsigned long age_jiffies = secs_to_jiffies(age_secs);
+	unsigned long age_jiffies = msecs_to_jiffies(age_secs * 1000);
 
 	spin_lock_bh(&rdev->bss_lock);
 	list_for_each_entry(bss, &rdev->bss_list, list)
