--- a/nss_core.c
+++ b/nss_core.c
@@ -477,7 +477,7 @@ static uint32_t nss_soc_mem_info(void)
 		goto err_use_default_memsize;
 	}
 
-	nss_info_always("NSS DDR size is 0x%x\n", (uint32_t) resource_size(&r));
+	nss_info("NSS DDR size is 0x%x\n", (uint32_t) resource_size(&r));
 
 	return resource_size(&r);
 
--- a/nss_hal/ipq50xx/nss_hal_pvt.c
+++ b/nss_hal/ipq50xx/nss_hal_pvt.c
@@ -449,15 +449,15 @@ static int __nss_hal_clock_configure(str
 		}
 	}
 
-	nss_info_always("Supported Frequencies - ");
+	nss_info("Supported Frequencies - ");
 	for (i = 0; i < NSS_FREQ_MAX_SCALE; i++) {
 		switch (nss_runtime_samples.freq_scale[i].frequency) {
 		case NSS_FREQ_850:
-			nss_info_always("850 MHz ");
+			nss_info("850 MHz ");
 			break;
 
 		case NSS_FREQ_1000:
-			nss_info_always("1 GHz ");
+			nss_info("1 GHz ");
 			break;
 
 		case NSS_FREQ_SCALE_NA:
@@ -468,7 +468,7 @@ static int __nss_hal_clock_configure(str
 			return -EFAULT;
 		}
 	}
-	nss_info_always("\n");
+	nss_info("\n");
 
 	/*
 	 * Set values only once for core0. Grab the proper clock.
--- a/nss_hal/ipq60xx/nss_hal_pvt.c
+++ b/nss_hal/ipq60xx/nss_hal_pvt.c
@@ -532,22 +532,22 @@ static int __nss_hal_clock_configure(str
 		}
 	}
 
-	nss_info_always("Supported Frequencies - ");
+	nss_info("Supported Frequencies - ");
 	for (i = 0; i < NSS_FREQ_MAX_SCALE; i++) {
 		if (nss_runtime_samples.freq_scale[i].frequency == NSS_FREQ_187) {
-			nss_info_always("187.2 MHz ");
+			nss_info("187.2 MHz ");
 		} else if (nss_runtime_samples.freq_scale[i].frequency == NSS_FREQ_748) {
-			nss_info_always("748.8 MHz ");
+			nss_info("748.8 MHz ");
 		} else if (nss_runtime_samples.freq_scale[i].frequency == NSS_FREQ_1497) {
-			nss_info_always("1.4976 GHz ");
+			nss_info("1.4976 GHz ");
 		} else if (nss_runtime_samples.freq_scale[i].frequency == NSS_FREQ_1689) {
-			nss_info_always("1.6896 GHz ");
+			nss_info("1.6896 GHz ");
 		} else {
 			nss_info_always("Error\nNo Table/Invalid Frequency Found\n");
 			return -EFAULT;
 		}
 	}
-	nss_info_always("\n");
+	nss_info("\n");
 
 	/*
 	 * Set values only once for core0. Grab the proper clock.
--- a/nss_hal/ipq807x/nss_hal_pvt.c
+++ b/nss_hal/ipq807x/nss_hal_pvt.c
@@ -571,22 +571,22 @@ static int __nss_hal_clock_configure(str
 		}
 	}
 
-	nss_info_always("Supported Frequencies - ");
+	nss_info("Supported Frequencies - ");
 	for (i = 0; i < NSS_FREQ_MAX_SCALE; i++) {
 		if (nss_runtime_samples.freq_scale[i].frequency == NSS_FREQ_187) {
-			nss_info_always("187.2 MHz ");
+			nss_info("187.2 MHz ");
 		} else if (nss_runtime_samples.freq_scale[i].frequency == NSS_FREQ_748) {
-			nss_info_always("748.8 MHz ");
+			nss_info("748.8 MHz ");
 		} else if (nss_runtime_samples.freq_scale[i].frequency == NSS_FREQ_1497) {
-			nss_info_always("1.4976 GHz ");
+			nss_info("1.4976 GHz ");
 		} else if (nss_runtime_samples.freq_scale[i].frequency == NSS_FREQ_1689) {
-			nss_info_always("1.6896 GHz ");
+			nss_info("1.6896 GHz ");
 		} else {
 			nss_info_always("Error\nNo Table/Invalid Frequency Found\n");
 			return -EFAULT;
 		}
 	}
-	nss_info_always("\n");
+	nss_info("\n");
 
 	/*
 	 * Set values only once for core0. Grab the proper clock.
