--- a/qmi_wwan_f.c
+++ b/qmi_wwan_f.c
@@ -712,7 +712,11 @@ static struct rtnl_link_stats64 *_rmnet_
         stats64 = per_cpu_ptr(dev->stats64, cpu);
 
         do {
+#if (LINUX_VERSION_CODE < KERNEL_VERSION( 6,6,0 ))
             start = u64_stats_fetch_begin_irq(&stats64->syncp);
+#else
+            start = u64_stats_fetch_begin(&stats64->syncp);
+#endif
 #if LINUX_VERSION_CODE < KERNEL_VERSION(5,18,0)
             rx_packets = stats64->rx_packets;
             rx_bytes = stats64->rx_bytes;
@@ -724,7 +728,11 @@ static struct rtnl_link_stats64 *_rmnet_
 	    tx_packets = u64_stats_read(&stats64->tx_packets);
 	    tx_bytes = u64_stats_read(&stats64->tx_bytes);
 #endif
+#if (LINUX_VERSION_CODE < KERNEL_VERSION( 6,6,0 ))
         } while (u64_stats_fetch_retry_irq(&stats64->syncp, start));
+#else
+        } while (u64_stats_fetch_retry(&stats64->syncp, start));
+#endif
 
         stats->rx_packets += rx_packets;
         stats->rx_bytes += rx_bytes;
