--- a/drivers/mtd/chips/cfi_cmdset_0002.c
+++ b/drivers/mtd/chips/cfi_cmdset_0002.c
@@ -1647,7 +1647,7 @@ static int __xipram do_write_oneword(str
 		}
 
 		if (chip_good(map, adr, &datum))
-			break;
+			goto enable_xip;
 
 		/* Latency issues. Drop the lock, wait a while and retry */
 		UDELAY(map, chip, adr, 1);
@@ -1664,6 +1664,8 @@ static int __xipram do_write_oneword(str
 			goto retry;
 		}
 	}
+
+ enable_xip:
 	xip_enable(map, chip, adr);
  op_done:
 	if (mode == FL_OTP_WRITE)
@@ -2246,7 +2248,6 @@ static int cfi_amdstd_panic_write(struct
 	return 0;
 }
 
-
 /*
  * Handle devices with one erase region, that only implement
  * the chip erase command.
@@ -2315,7 +2316,7 @@ static int __xipram do_erase_chip(struct
 		}
 
 		if (chip_ready(map, adr, &datum))
-			break;
+			goto op_done;
 
 		if (time_after(jiffies, timeo)) {
 			printk(KERN_WARNING "MTD %s(): software timeout\n",
@@ -2339,6 +2340,7 @@ static int __xipram do_erase_chip(struct
 		}
 	}
 
+ op_done:
 	chip->state = FL_READY;
 	xip_enable(map, chip, adr);
 	DISABLE_VPP(map);
@@ -2413,7 +2415,7 @@ static int __xipram do_erase_oneblock(st
 
 		if (chip_ready(map, adr, &datum)) {
 			xip_enable(map, chip, adr);
-			break;
+			goto op_done;
 		}
 
 		if (time_after(jiffies, timeo)) {
@@ -2439,6 +2441,7 @@ static int __xipram do_erase_oneblock(st
 		}
 	}
 
+ op_done:
 	chip->state = FL_READY;
 	DISABLE_VPP(map);
 	put_chip(map, chip, adr);
